lib/Service/ImageService.php

```diff
@@ -0,0 +1,676 @@
+	/**
+	 * @var IPreview
+	 */
+	private $previewManager;
+
+	public function __construct(IRootFolder $rootFolder,
+								LoggerInterface $logger,
+								ShareManager $shareManager,
+								IPreview $previewManager,
+								IClientService $clientService) {
+		$this->rootFolder = $rootFolder;
+		$this->shareManager = $shareManager;
+		$this->clientService = $clientService;
+		$this->logger = $logger;
+		$this->previewManager = $previewManager;
+	}
+
+	/**
+	 * Get image content or preview from file id
+	 * @param int $documentId
+	 * @param string $imageFileName
+	 * @param string $userId
+	 * @return File|\OCP\Files\Node|ISimpleFile|null
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	public function getImage(int $documentId, string $imageFileName, string $userId) {
+		$textFile = $this->getTextFile($documentId, $userId);
+		return $textFile === null ? null : $this->getImagePreview($imageFileName, $textFile);
+	}
+
+	/**
+	 * Get image content or preview from file id in public context
+	 * @param int $documentId
+	 * @param string $imageFileName
+	 * @param string $shareToken
+	 * @return File|\OCP\Files\Node|ISimpleFile|null
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	public function getImagePublic(int $documentId, string $imageFileName, string $shareToken) {
+		$textFile = $this->getTextFilePublic($documentId, $shareToken);
+		return $textFile === null ? null : $this->getImagePreview($imageFileName, $textFile);
+	}
+
+	/**
+	 * @param string $imageFileName
+	 * @param File $textFile
+	 * @return File|\OCP\Files\Node|ISimpleFile|null
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	private function getImagePreview(string $imageFileName, File $textFile) {
+		$attachmentFolder = $this->getOrCreateAttachmentDirectoryForFile($textFile);
+		if ($attachmentFolder !== null) {
+			try {
+				$imageFile = $attachmentFolder->get($imageFileName);
+			} catch (NotFoundException $e) {
+				return null;
+			}
+			if ($imageFile instanceof File) {
+				if ($this->previewManager->isMimeSupported($imageFile->getMimeType())) {
+					return $this->previewManager->getPreview($imageFile, 1024, 1024);
+				} else {
+					return $imageFile;
+				}
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Save an uploaded image in the attachment folder
+	 *
+	 * @param int $documentId
+	 * @param string $newFileName
+	 * @param string $newFileContent
+	 * @param string $userId
+	 * @return array
+	 * @throws NotFoundException
+	 * @throws \OCP\Files\InvalidPathException
+	 * @throws \OCP\Files\NotPermittedException
+	 */
+	public function uploadImage(int $documentId, string $newFileName, string $newFileContent, string $userId): array {
+		$textFile = $this->getTextFile($documentId, $userId);
+		if (!$textFile->isUpdateable()) {
```

Comment:
Better to check for $textFile to prevent null pointer exception:

```suggestion
		if (!$textFile || !$textFile->isUpdateable()) {
```

Or even better, throw a `NotFoundException` if empty in `$this->getTextFile()`?

Reply:
Yep. `ImageController::getImage()` now catches the exceptions.
`ImageService::getTextFile()` and `ImageService::getTextFilePublic()` now throw a `NotFoundException`.